### DISCUSSION TITLE
1 MILLION EPOCHS 😱😱😱

### DISCUSSION THREAD

**Discussion by furquanSal** [mrdbourke/pytorch-deep-learning#745] [Show and tell]
Hey Daniel and everyone, I hope you are all doing well!!!

I am Mohammed Furquan, a final-year AIML Engineering student from India. I have been watching Daniel's PyTorch course for a few days and I am enjoying it. I just now completed training my model. I just wanted to share that I ran my model with 1 Million epochs and I was wondering what could that turn out to be. Would love to share the results.....
![1M-epochs](https://github.com/mrdbourke/pytorch-deep-learning/assets/109865936/7b130d88-22dd-4b8f-b95d-f7a41f0610b3)
![1M-graph](https://github.com/mrdbourke/pytorch-deep-learning/assets/109865936/8132f44d-d201-4de3-bca4-0f4958341960)
![1M-results](https://github.com/mrdbourke/pytorch-deep-learning/assets/109865936/2df98297-561b-42f5-bf71-35021608c8e3)

**Comment by bhuvanmdev** [comment]
it's true that most of the times for a given data, as you increase the number of epochs, the accuracy might increase (sometimes it does'nt and sometimes even starts decreasing) but the reason why we usually don't do this is..
* This easily leads to **overfitting**
* Improvement in results is not worth the **computation required**.  
Instead of increasing the epochs, we need to use **regularization techniques** , **try different architecture** , **normalize the data** etc..

Its fine to do this as a learning experience, but it's not a good practice in general.

**Reply by furquanSal** [reply]
Yup! It's impractical when building a complex Machine Learning model as it will take days to complete. I just wanted to experience this and share it with you all merely for fun.

Thanks for the advice anyway. Have a good day! 😉

**Comment by heebahsaleem** [comment]
Hi,
I am training my model for traffic light signal detection using openAI OAK-1 camera, which will be used in real time autonomous vehicle. 
The model I have used is MobileNetV2

So, it is advised from my senior, to train the model with 1million epochs. 
Q1. Do you suggest me to train it with 1 million epochs?
Q2. Are there any chances of overfitting? 
Q3. How much should i set Early stopping patience?

Any help or advise would be appreciated^^

**Reply by bhuvanmdev** [reply]
I have no personal experience training any kind of mobilenet models,  but in general bigger models are prone to overfitting and smaller models are prone to underfitting and also the total number of epochs or any hyper parameter for that matter can vary depending on the
- size of architecture
- size and type of dataset
- optimization used etc...

To check whether your model is overfitting or not you can use.
- early stop helper functions
- use a train and validation data set to check for the train_loss and valid_loss for each epoch(or at equal intervals) and plot the curves.
Check for the distance between both the curves. 
Too much distance == underfitting
Too less distance == overfitting

And importantly when you experiment with the model to find suitable epoch, make sure all other hyper parameters are constant, this also includes the **seed value**(i.e constant randomness).

I hope this helps :)
